File: calculator.cpp
#include "calculator.h"
using namespace std;

double calcAvg(course_inf c){
    double quizavg = 0;
    double examavg = 0;
    double homeworkavg = 0;
    double assignmentavg = 0;
    int quiznum = 0;
    int examnum = 0;
    int homeworknum = 0;
    int assignmentnum = 0;

    for(auto it = c.assignments.begin(); it != c.assignments.end(); it++){
        if((*it) -> type == "quiz"){
            quizavg += (*it)->grade;
            quiznum += 1;
        }
        else if((*it) -> type == "exam"){
            examavg += (*it)->grade;
            examnum += 1;
        }
        else if((*it) -> type == "homework"){
            homeworkavg += (*it)->grade;
            homeworknum += 1;
        }
        else if((*it) -> type == "assignment"){
            assignmentavg += (*it)->grade;
            assignmentnum += 1;
        }
    }
    quizavg /= quiznum;
    // cout << "quizavg: " << quizavg << endl;
    examavg /= examnum;
    // cout << "examnum: " << examavg << endl;
    homeworkavg /= homeworknum;
    // cout << "homeworkavg: " << homeworkavg << endl;
    assignmentavg /= assignmentnum;
    // cout << "assignmentavg: " << assignmentavg << endl;

    double totavg = quizavg * c.quizWeight + examavg * c.examWeight + homeworkavg * c.homeworkWeight + assignmentavg * c.assignmentWeight;
    return totavg;
}

double calcNeededQuizAvg(double grade, course_inf c){
    double examavg = 0;
    double homeworkavg = 0;
    double assignmentavg = 0;
    int examnum = 0;
    int homeworknum = 0;
    int assignmentnum = 0;

    for(auto it = c.assignments.begin(); it != c.assignments.end(); it++){
        if((*it) -> type == "exam"){
            examavg += (*it)->grade;
            examnum += 1;
        }
        else if((*it) -> type == "homework"){
            homeworkavg += (*it)->grade;
            homeworknum += 1;
        }
        else if((*it) -> type == "assignment"){
            assignmentavg += (*it)->grade;
            assignmentnum += 1;
        }
    }
    examavg /= examnum;
    // cout << "examnum: " << examavg << endl;
    homeworkavg /= homeworknum;
    // cout << "homeworkavg: " << homeworkavg << endl;
    assignmentavg /= assignmentnum;
    // cout << "assignmentavg: " << assignmentavg << endl;

    double neededavg = grade - (examavg * c.examWeight + homeworkavg * c.homeworkWeight + assignmentavg * c.assignmentWeight);
    // cout << "Needed Total: " << neededavg << endl;
    neededavg /= c.quizWeight;
    return neededavg;
}

double calcNeededExamAvg(double grade, course_inf c){
    double quizavg = 0;
    double homeworkavg = 0;
    double assignmentavg = 0;
    int quiznum = 0;
    int homeworknum = 0;
    int assignmentnum = 0;

    for(auto it = c.assignments.begin(); it != c.assignments.end(); it++){
        if((*it) -> type == "quiz"){
            quizavg += (*it)->grade;
            quiznum += 1;
        }
        else if((*it) -> type == "homework"){
            homeworkavg += (*it)->grade;
            homeworknum += 1;
        }
        else if((*it) -> type == "assignment"){
            assignmentavg += (*it)->grade;
            assignmentnum += 1;
        }
    }
    quizavg /= quiznum;
    homeworkavg /= homeworknum;
    assignmentavg /= assignmentnum;

    double neededavg = grade - (quizavg * c.quizWeight + homeworkavg * c.homeworkWeight + assignmentavg * c.assignmentWeight);
    neededavg /= c.examWeight;

    return neededavg;
}

double calcNeededHomeworkAvg(double grade, course_inf c){
    double quizavg = 0;
    double examavg = 0;
    double assignmentavg = 0;
    int quiznum = 0;
    int examnum = 0;
    int assignmentnum = 0;

    for(auto it = c.assignments.begin(); it != c.assignments.end(); it++){
        if((*it) -> type == "quiz"){
            quizavg += (*it)->grade;
            quiznum += 1;
        }
        else if((*it) -> type == "exam"){
            examavg += (*it)->grade;
            examnum += 1;
        }
        else if((*it) -> type == "assignment"){
            assignmentavg += (*it)->grade;
            assignmentnum += 1;
        }
    }
    quizavg /= quiznum;
    examavg /= examnum;
    assignmentavg /= assignmentnum;

    double neededavg = grade - (quizavg * c.quizWeight + examavg * c.examWeight + assignmentavg * c.assignmentWeight);
    neededavg /= c.homeworkWeight;

    return neededavg;
}

double calcNeededAssignmentAvg(double grade, course_inf c){
    double quizavg = 0;
    double examavg = 0;
    double homeworkavg = 0;
    int quiznum = 0;
    int examnum = 0;
    int homeworknum = 0;

    for(auto it = c.assignments.begin(); it != c.assignments.end(); it++){
        if((*it) -> type == "quiz"){
            quizavg += (*it)->grade;
            quiznum += 1;
        }
        else if((*it) -> type == "exam"){
            examavg += (*it)->grade;
            examnum += 1;
        }
        else if((*it) -> type == "homework"){
            homeworkavg += (*it)->grade;
            homeworknum += 1;
        }
    }
    quizavg /= quiznum;
    examavg /= examnum;
    homeworkavg /= homeworknum;

    double neededavg = grade - (quizavg * c.quizWeight + examavg * c.examWeight + homeworkavg * c.homeworkWeight);
    neededavg /= c.assignmentWeight;

    return neededavg;
}

File: CALCULATOR.py
from COURSE_INF import CourseInf, AssignmentInf

def calc_avg(c, average_type):
    quiz_avg = 0
    exam_avg = 0
    homework_avg = 0
    assignment_avg = 0
    quiz_num = 0
    exam_num = 0
    homework_num = 0
    assignment_num = 0

    for assignment in c.assignments:
        if assignment.type == "quiz":
            quiz_avg += assignment.grade
            quiz_num += 1
        elif assignment.type == "exam":
            exam_avg += assignment.grade
            exam_num += 1
        elif assignment.type == "homework":
            homework_avg += assignment.grade
            homework_num += 1
        elif assignment.type == "assignment":
            assignment_avg += assignment.grade
            assignment_num += 1

    quiz_avg /= quiz_num
    exam_avg /= exam_num
    homework_avg /= homework_num
    assignment_avg /= assignment_num

    total_avg = (quiz_avg * c.quizWeight + exam_avg * c.examWeight +
                 homework_avg * c.homeworkWeight + assignment_avg * c.assignmentWeight)
    
    if average_type == "quiz":
        return quiz_avg
    elif average_type == "assignment":
        return assignment_avg
    elif average_type == "exam":
        return exam_avg
    elif average_type == "homework":
        return homework_avg
    elif average_type == "class":
        return total_avg


def calc_needed_quiz_avg(grade, c):
    exam_avg = 0
    homework_avg = 0
    assignment_avg = 0
    exam_num = 0
    homework_num = 0
    assignment_num = 0

    for assignment in c.assignments:
        if assignment.type == "exam":
            exam_avg += assignment.grade
            exam_num += 1
        elif assignment.type == "homework":
            homework_avg += assignment.grade
            homework_num += 1
        elif assignment.type == "assignment":
            assignment_avg += assignment.grade
            assignment_num += 1

    exam_avg /= exam_num
    homework_avg /= homework_num
    assignment_avg /= assignment_num

    needed_avg = grade - (exam_avg * c.examWeight + homework_avg * c.homeworkWeight + assignment_avg * c.assignmentWeight)
    needed_avg /= c.quizWeight
    return needed_avg


def calc_needed_exam_avg(grade, c):
    quiz_avg = 0
    homework_avg = 0
    assignment_avg = 0
    quiz_num = 0
    homework_num = 0
    assignment_num = 0

    for assignment in c.assignments:
        if assignment.type == "quiz":
            quiz_avg += assignment.grade
            quiz_num += 1
        elif assignment.type == "homework":
            homework_avg += assignment.grade
            homework_num += 1
        elif assignment.type == "assignment":
            assignment_avg += assignment.grade
            assignment_num += 1

    quiz_avg /= quiz_num
    homework_avg /= homework_num
    assignment_avg /= assignment_num

    needed_avg = grade - (quiz_avg * c.quizWeight + homework_avg * c.homeworkWeight + assignment_avg * c.assignmentWeight)
    needed_avg /= c.examWeight
    return needed_avg


def calc_needed_homework_avg(grade, c):
    quiz_avg = 0
    exam_avg = 0
    assignment_avg = 0
    quiz_num = 0
    exam_num = 0
    assignment_num = 0

    for assignment in c.assignments:
        if assignment.type == "quiz":
            quiz_avg += assignment.grade
            quiz_num += 1
        elif assignment.type == "exam":
            exam_avg += assignment.grade
            exam_num += 1
        elif assignment.type == "assignment":
            assignment_avg += assignment.grade
            assignment_num += 1

    quiz_avg /= quiz_num
    exam_avg /= exam_num
    assignment_avg /= assignment_num

    needed_avg = grade - (quiz_avg * c.quizWeight + exam_avg * c.examWeight + assignment_avg * c.assignmentWeight)
    needed_avg /= c.homeworkWeight
    return needed_avg


def calc_needed_assignment_avg(grade, c):
    quiz_avg = 0
    exam_avg = 0
    homework_avg = 0
    quiz_num = 0
    exam_num = 0
    homework_num = 0

    for assignment in c.assignments:
        if assignment.type == "quiz":
            quiz_avg += assignment.grade
            quiz_num += 1
        elif assignment.type == "exam":
            exam_avg += assignment.grade
            exam_num += 1
        elif assignment.type == "homework":
            homework_avg += assignment.grade
            homework_num += 1

    quiz_avg /= quiz_num
    exam_avg /= exam_num
    homework_avg /= homework_num

    needed_avg = grade - (quiz_avg * c.quizWeight + exam_avg * c.examWeight + homework_avg * c.homeworkWeight)
    needed_avg /= c.assignmentWeight
    return needed_avg
